plugins/browser-plugin-youtube-tracking/src/index.ts

```diff
@@ -0,0 +1,231 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import { UrlParameters, trackingOptionsParser } from './helperFunctions';
+import { YTError, YTPlayerEvent, YTState, YTStateEvent } from './constants';
+import { EventData, MediaTrackingOptions, TrackedPlayer, TrackingOptions } from './types';
+import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
+import { buildSelfDescribingEvent, CommonEventProperties, Logger, SelfDescribingJson } from '@snowplow/tracker-core';
+import { SnowplowEvent } from './snowplowEvents';
+import { MediaPlayerEvent } from './contexts';
+import { buildYouTubeEvent } from './buildYouTubeEvent';
+import { YTEvent } from './youtubeEvents';
+
+declare global {
+  interface Window {
+    onYouTubeIframeAPIReady?: Function;
+  }
+}
+
+let LOG: Logger;
+export const _trackers: Record<string, BrowserTracker> = {};
+let tracked_players: Record<string, TrackedPlayer> = {};
+let iframeApiHasLoaded = false;
+let trackingQueue: Array<TrackingOptions> = [];
+
+export function YouTubeTrackingPlugin(): BrowserPlugin {
+  return {
+    activateBrowserPlugin: (tracker: BrowserTracker) => {
+      _trackers[tracker.id] = tracker;
+    },
+    logger: (logger) => {
+      LOG = logger;
+    },
+  };
+}
+
+function trackEvent(
+  event: SelfDescribingJson<MediaPlayerEvent> & CommonEventProperties,
+  trackers: Array<string> = Object.keys(_trackers)
+): void {
+  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
+    t.core.track(buildSelfDescribingEvent({ event }), event.context, event.timestamp);
+  });
+}
+
+function enableYouTubeIframeAPI() {
+  const tag: HTMLScriptElement = document.createElement('script');
+  tag.src = 'https://www.youtube.com/iframe_api';
+  const firstScriptTag = document.getElementsByTagName('script')[0];
+  firstScriptTag.parentNode!.insertBefore(tag, firstScriptTag);
+}
+
+export function enableYouTubeTracking(args: { id: string; options?: MediaTrackingOptions }) {
+  const conf: TrackingOptions = trackingOptionsParser(args.id, args.options);
+  const el: HTMLIFrameElement = document.getElementById(args.id) as HTMLIFrameElement;
+  if (!el) {
+    LOG.error('Cannot find YouTube IFrame element');
+    return;
+  }
+  conf.urlParameters = new UrlParameters(el.src);
+  el.src = conf.urlParameters.getUrl();
+  enableYouTubeIframeAPI();
```

Comment:
Similar to my earlier comment, what if a user has already added this to their website? Maybe they use another tool that needs the API on their page. Would we end up loading it twice, with the later load not being required.